examples/stable_diffusion_v2/lora/lora.py

```diff
@@ -0,0 +1,234 @@
+            compute_dtype=mstype.float16,
+            **kwargs
+    ):
+        super(LoRADense, self).__init__(in_channels, out_channels, **kwargs)
+
+        # Check params
+        self._check_num(lora_rank, lora_alpha, lora_dropout)
+        self._check_init(lora_a_init, lora_b_init, lora_rank)
+        self._check_type_of_data(param_init_type, compute_dtype)
+
+        # Define and initialize params
+        self.lora_rank = lora_rank
+        self.lora_alpha = lora_alpha
+        if is_version_ge(ms.__version__, '2.0.0'):
+            self.lora_dropout = nn.Dropout(p=lora_dropout)
+        else:
+            self.lora_dropout = nn.Dropout(keep_prob=1 - lora_dropout)
+        self.tk_delta_lora_a = Parameter(
+            initializer(lora_a_init, [lora_rank, in_channels], param_init_type),
+            name='tk_delta_lora_A')
+        self.tk_delta_lora_b = Parameter(initializer(lora_b_init, [out_channels, lora_rank], param_init_type),
+                                         name='tk_delta_lora_B')
+        self.scaling = self.lora_alpha / self.lora_rank
+
+        # Calculation utils
+        self.mul = ops.Mul()
+        self.add = ops.Add()
+        self.cast = ops.Cast()
+        self.dtype = compute_dtype
+        self.lora_a_matmul = P.MatMul(transpose_b=True)
+        self.lora_b_matmul = P.MatMul(transpose_b=True)
+
+    def construct(self, input_tensor):
+        # Data type operation
+        ori_dtype = F.dtype(input_tensor)
+        input_tensor = self.cast(input_tensor, self.dtype)
+        weight = self.cast(self.weight, self.dtype)
+        lora_a = self.cast(self.tk_delta_lora_a, self.dtype)
+        lora_b = self.cast(self.tk_delta_lora_b, self.dtype)
+        scaling = self.cast(self.scaling, self.dtype)
+
+        # Shape operations
+        x_shape = self.shape_op(input_tensor)
+        nn.layer.basic.check_dense_input_shape(x_shape, self.cls_name)
+        input_tensor = self.reshape(input_tensor, (-1, x_shape[-1]))
+
+        # Dense result
+        dense_result = self.matmul(input_tensor, weight)
+        if self.has_bias:
+            bias = self.cast(self.bias, self.dtype)
+            dense_result = self.bias_add(dense_result, bias)
+
+        # LoRA result
+        input_tensor = self.lora_dropout(input_tensor)
+        input_tensor = self.lora_a_matmul(input_tensor, lora_a)
+        input_tensor = self.lora_b_matmul(input_tensor, lora_b)
+        input_tensor = self.mul(input_tensor, scaling)
+
+        # Result addition and activation
+        dense_result = self.add(dense_result, input_tensor)
+        if self.activation_flag:
+            dense_result = self.activation(dense_result)
+
+        # Shape restore
+        if len(x_shape) != 2:
+            out_shape = x_shape[:-1] + (-1,)
+            dense_result = self.reshape(dense_result, out_shape)
+        dense_result = self.cast(dense_result, ori_dtype)
+
+        return dense_result
+
+    def shard(self, strategy_org_dense_matmul=None,
```

Comment:
please delete this part since it is not used in current data parallel mode.

Reply:
ok